internal/gossiphttp/transport.go

```diff
@@ -0,0 +1,554 @@
+		}
+	}()
+
+	<-ctx.Done()
+}
+
+// Handler returns the base HTTP route and handler for the Transport.
+func (t *Transport) Handler() (route string, handler http.Handler) {
+	mux := http.NewServeMux()
+	mux.Handle(messageEndpoint, http.HandlerFunc(t.handleMessage))
+	mux.Handle(streamEndpoint, http.HandlerFunc(t.handleStream))
+	return baseRoute, mux
+}
+
+// handleMessage is used for single-packet communication.
+func (t *Transport) handleMessage(w http.ResponseWriter, r *http.Request) {
+	if r.ProtoMajor != 2 {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	if r.Header.Get(contentTypeHeader) != ckitContentType {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	var (
+		recvTime   = time.Now()
+		remoteAddr = parseRemoteAddr(r.RemoteAddr)
+	)
+
+	// Read each message until the request body has been fully consumed. Each
+	// message is converted into a single packet.
+	for {
+		msg, err := readMessage(r.Body)
+		if errors.Is(err, io.EOF) {
+			break
+		} else if err != nil {
+			level.Warn(t.log).Log("msg", "error reading packet from peer", "err", err)
+			break
+		}
+
+		t.metrics.packetRxTotal.Inc()
+		t.metrics.packetRxBytesTotal.Add(float64(len(msg)))
+
+		// Enqueue the packet to be processed in the background. This allows
+		// HTTP calls to have as low of a latency as possible to help keep
+		// things moving along.
+		t.inPacketQueue.Enqueue(&memberlist.Packet{
+			Buf:       msg,
+			From:      remoteAddr,
+			Timestamp: recvTime,
+		})
+	}
+
+	w.WriteHeader(http.StatusOK)
+}
+
+// parseRemoteAddr parses a ip:port string into a net.Addr. If the addr cannot
+// be parsed, a default implementation for an "unknown" net.Addr is returned.
+func parseRemoteAddr(addr string) net.Addr {
+	remoteHost, remoteService, err := net.SplitHostPort(addr)
+	if err != nil {
+		return unknownAddr{}
+	}
+
+	remoteIP := net.ParseIP(remoteHost)
+	if remoteIP == nil {
+		return unknownAddr{}
+	}
+
+	remotePort, err := net.LookupPort("tcp", remoteService)
+	if err != nil {
+		return unknownAddr{}
+	}
+
+	return &net.TCPAddr{
+		IP:   remoteIP,
+		Port: remotePort,
+	}
+}
+
+// handleStream is used for streaming bidirectional communication.
+func (t *Transport) handleStream(w http.ResponseWriter, r *http.Request) {
+	if r.ProtoMajor != 2 {
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+	flusher, ok := w.(http.Flusher)
+	if !ok {
+		w.WriteHeader(http.StatusHTTPVersionNotSupported)
+		return
+	}
+
+	if r.Header.Get(contentTypeHeader) != ckitContentType {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	w.WriteHeader(http.StatusOK)
+	flusher.Flush()
+
+	waitClosed := make(chan struct{})
+
+	var readMut sync.Mutex
+	readCnd := sync.NewCond(&readMut)
+
+	t.metrics.openStreams.Inc()
+
+	packetsClient := &http2Stream{
+		t: t,
+		r: r.Body,
+		w: &flushWriter{w: w, f: flusher},
+	}
+
+	conn := &packetsClientConn{
+		cli: packetsClient,
+		onClose: func() {
+			t.metrics.openStreams.Dec()
+			close(waitClosed)
+		},
```

Comment:
(Given the suggestion above)

```suggestion
		onClose: func() { close(waitClosed) },
```